docs/ldml/tr35-keyboards.md

```diff
@@ -2908,6 +2908,196 @@ This attribute specifies the expected resultant text in a document after process
+            const segment    = input.substring(segmentStart, segmentEnd);
+            const segmentNfd = nfdMarkersSegment(segment);
+            output.append(segmentNfd);
+            segmentStart = segmentEnd;
+        }
+    } while(segmentEnd != input.end);
+
+    return output;
+}
+
+/**
+ * @param input segment text
+ * @returns normalized output segment
+ */
+function nfdMarkersSegment(input /* string with markers */) {
+    const noMarkers    = removeMarkers(input);
+    const markerMap    = parseMarkers(input);
+    const noMarkersNfd = normalizeNFD(noMarkers);
+    if (markerMap.empty()) {
+        // no markers in this segment.
+        return noMarkersNfd;
+    } else if(noMarkersNfd == noMarkers) {
+        // no change in this segment
+        return input;
+    } else {
+        return addBackMarkers(noMarkersNfd, markerMap);
+    }
+}
+
+// represents a removed marker, to be re-added
+interface MarkerEntry {
+    glue: string;      // 'glue' character
+    marker: Marker;  // which marker
+    processed: boolean = false; // track whether this was processed or not
+    eot: boolean = false; // track whether end of text or not
+}
+
+/** parse the string for markers */
+function parseMarkers(s: string): MarkerEntry[] {
+    // Note: parseMarkers() and removeMarkers() (not shown) could be the same function.
+    // queue of markers found
+    let lastMarkers: Marker[] = [];
+    // return value
+    let outputMap: MarkerEntry[] = [];
+
+
+    // iterate string by codepoints
+    for (i = s.begin; i< s.end;) {
+        if (markerAt(i)) {
+            // found a marker - add it to the list
+            lastMarkers.add(markerAt(i));
+            i += /* marker length */;
+        } else {
+            // From 'i' find the glue char for all markers found so far
+            // (in order)
+            // use first char of decomposed string
+            const glue = normalizeNFD(i).codePointAt(0);
+            for (m of markers) {
+                map.add({marker: m, glue: glue});
+            }
+            markers.clear();
+
+            i++; // move to next codepoint
+        }
+    }
+
+    // ran out of codepoints.
+    // any remaining markers are EOT
+    for (m of markers) {
+        map.add({marker: m, eot: true});
+    }
+}
+
+/**
+ * re-add markers to input. Mutates markerMap.
+ * @param input normalized string to re-add markers
+ * @param markerMap list of glued markers. will be mutated
+ */
+function addBackMarkers(input : string, map: MarkerEntry[]) : stringWithMarkers {
+    // quick check
+    if(map.empty()) return input;
+
+    // output string
+    let out = '';
+
+    // first, add all of the EOT entries
+    while (!map.empty() && map[map.length - 1].eot) {
+        const m = map.popEnd(); // remove the last entry
+        out = m.marker + out; // prepend the marker
+    }
+
+    // now, process input codepoints from end to beginning
+    for (i=input.end;i>=input.begin;i--) {
```

Comment:
Since this code's been spun off, here's a link to my old comment re: this section - https://github.com/unicode-org/cldr/pull/3445#discussion_r1468993302